src/kagglehub/gcs_upload.py

```diff
@@ -138,51 +147,110 @@ def _upload_blob(file_path: str, model_type: str) -> str:
-
-    with Pool() as pool:
-        sizes = pool.map(zip_file, args)
-    return sizes
-
-
-def upload_files(source_path: str, model_type: str) -> List[str]:
-    source_path_obj = Path(source_path)
-    with TemporaryDirectory() as temp_dir:
-        temp_dir_path = Path(temp_dir)
-        total_size = 0
-
-        if source_path_obj.is_dir():
-            for file_path in source_path_obj.rglob("*"):
-                if file_path.is_file():
-                    total_size += file_path.stat().st_size
-        elif source_path_obj.is_file():
-            total_size = source_path_obj.stat().st_size
-        else:
-            path_error_message = "The source path does not point to a valid file or directory."
-            raise ValueError(path_error_message)
-
-        with tqdm(total=total_size, desc="Zipping", unit="B", unit_scale=True, unit_divisor=1024) as pbar:
-            if source_path_obj.is_dir():
-                zip_path = temp_dir_path / "archive.zip"
-                sizes = zip_files(source_path_obj, zip_path)
-                for size in sizes:
-                    pbar.update(size)
-                upload_path = str(zip_path)
-            elif source_path_obj.is_file():
-                temp_file_path = temp_dir_path / source_path_obj.name
-                shutil.copy(source_path_obj, temp_file_path)
-                pbar.update(temp_file_path.stat().st_size)
-                upload_path = str(temp_file_path)
-
-        return [token for token in [_upload_blob(upload_path, model_type)] if token]
+def upload_files_and_directories(
+    folder: str, model_type: str, quiet: bool = False  # noqa: FBT002, FBT001
+) -> UploadDirectoryInfo:
+    # Count the total number of files
+    file_count = 0
+    for _, _, files in os.walk(folder):
+        file_count += len(files)
+
+    if file_count > MAX_FILES_TO_UPLOAD:
+        if not quiet:
+            logger.info(f"More than {MAX_FILES_TO_UPLOAD} files detected, creating a zip archive...")
+
+        with TemporaryDirectory() as temp_dir:
+            zip_path = os.path.join(temp_dir, TEMP_ARCHIVE_FILE)
+            with zipfile.ZipFile(zip_path, "w") as zipf:
+                for root, _, files in os.walk(folder):
+                    for file in files:
+                        file_path = os.path.join(root, file)
+                        zipf.write(file_path, os.path.relpath(file_path, folder))
+
+            tokens = [
+                token
+                for token in [_upload_file_or_folder(temp_dir, TEMP_ARCHIVE_FILE, model_type, quiet)]
+                if token is not None
+            ]
+            return UploadDirectoryInfo(name="archive", files=tokens)
+
+    root_dict = UploadDirectoryInfo(name="root")
+    if os.path.isfile(folder):
+        # Directly upload the file if the path is a file
+        file_name = os.path.basename(folder)
+        token = _upload_file_or_folder(os.path.dirname(folder), file_name, model_type, quiet)
+        if token:
```

Comment:
what happens if token is falsy? What is the desired behavior?

Reply:
```
if upload_response.status_code in [200, 201]:
                    return response["token"]
                else:
                    upload_failed_exception = (
                        f"Upload failed with status code {upload_response.status_code}: {upload_response.text}"
                    )
                    raise BackendError(upload_failed_exception)
```

an exception will be raised upstream in upload_blob if a filed failed to upload. If the file was uploaded successfully however and we got an empty token somehow, this is where this check comes into play. Maybe, we should print to the user something, but not sure what. because the upload didnt fail (but we didnt get a token back) this scenario is unlikely.